libgedathon/docs/geda-snap-grid.1.in
.TH geda-snap-grid 1 "@DATE@" "gEDA Project" @VERSION@

.SH NAME
geda-snap-grid \- A gEDA-gaf Schematics and Symbol cleaning utility

.SH SYNOPSIS
.B geda-snap-grid
.RI  [ INFORMATION ] | [ OPTIONS ] [ DIRECTIVES ] [INPUT] [OUTPUT]

.SH DESCRIPTION
Round X-Y coordinates of attributes in schematic and symbol files to multibles
of 25 or a specified value.

By default, geda-snap-grid operates on both schematic and symbol files
unless file names are specified on the command line, either with the
--input <inputfile> option or just listing the files. For both schematic
and symbol files, the extension must be specified when files names are
specified.

.SH QUICK
.B usage: [Options] -i <inputfile> [[-o] <outputfile> ] [list of files]

Example: Snap all attribute coordinates values in all files in the current
         directory and in all subdirectories to the nearest multiple of 50
         and report modified values:

    geda-snap-grid.py --verbose -R -m 50

Example: Snap all attribute coordinates values in schematic files in the current
         directory appending "sym" and "sht" to the library search path rounding
         up to a multiple of 25:

    geda-snap-grid.py -a sht --sch

Example: Snap all attribute coordinates values in the file Motor_Control_sht_2.sch
         up to a multiple of 25 and save the file to Motor_Control_sht_2.fixed.sch,
         appending "sym" to the library search path if "./sym/" exist:

    geda-snap-grid -u -i Motor_Control_sht_2.sch -o Motor_Control_sht_2.fixed.sch

.SH INFORMATION
Use of the following options displays information related to geda-snap-grid
but does not inspect or alter any symbol files:
.IP "-h, --help"
Show information about the advanced usage of this command.
.IP "--usage"
Displays basic usage of this command.
.IP "-V, --version"
Displays software version information.

.SH OPTIONS
The following options can be used to modify the default behavior of geda-snap-grid

.IP "-d, --down"
"Force rounding down, the default is to round to nearest multible of moduli.
.IP "-u, --up"
Force rounding up, the default is to round to nearest multible of moduli.
.IP "-m #"
Round to moduli specified by #, default is 25.
.IP "-x"
Only round X coordinates, mutually exclusive with -y
.IP "-y"
Only round Y coordinates, mutually exclusive with -x

.SH DIRECTIVES
.IP "-v, --verbose"
Enable verbose mode. Displays additional information about what the program is doing.
.IP "-r, --recursive"
Process all symbol files in the current and all subordinated directories.
.IP "-a, --append <path>"
When processing schematic files this option can be used to specify addition directories to
be appended to the libraries search path. This option is not needed for directories with
the name "sym", if such a directory exist, the directory will be appended automatically.
.IP "-n, --no-path"
By default, if a directory exist with the name "sym", then the directory is appended to
the library search path when processing schematic files, this option inhibits that behavior.
This option is provided for advanced users and should not normally be used.
.IP "-c, --sch"
Directs geda-snap-grid to only process schematic files. This directive is ignored if a
file name or names are specified on the command line.
.IP "-s, --sym"
Directs geda-snap-grid to only process symbol files. This directive is ignored if a file
name or names are specified on the command line.
.IP "-i, --input <input file>"
Specify the name of the input file. Wild cards are accepted, the default is *.sym + *.sch.
.IP "-o, --output <output file>"
Specifies the name of the output file. When the output file name is not specified the input
file will be over written

.SH HOMEPAGE
http://www.geda-project.org/

.SH BUGS
You can report bugs at the Launchpad site of geda-snap-grid:
https://bugs.launchpad.net/geda/geda-snap-grid/+filebug

.SH AUTHOR
Wiley Edward Hill < wileyhill dot gmail dot com >

.SH SEE ALSO
.BR gsym-attribute (1) gsym-set-font-size (1) gsym-set-visibility (1)

.SH COPYRIGHT
.nf
Copyright \(co @YEAR@ Wiley Edward Hill.  License GPLv3+: GNU GPL
version 3.  Please see the `COPYING' file included with this
program for full details.
.PP
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

.


NEWS
==============================================
 'gschem and Friends' Electronic Design Suite
==============================================

Copyright (C) 1998-2017 gEDA Developers

This file documents important user-visible changes in gEDA/gaf.  For
more information, please consult the `ChangeLog' file.

Notable changes in gEDA/gaf gEDA 2.1.1.1 - 11/12/17 Rampage Edition

* Expanded support for internationalization (i18n).

* The build system now support using the --with-included-gettext option to
  configure the suite to build and use an in-source international library.

* Font-sizes for widgets on GedaToolbars can now be specified in RC files.

* The toolbar layout in gschem has been improved to better support labels.

* The visibility of toolbar tooltips can now be control using RC keywords.

* The Slot Editor Dialog now incorporates a spinner control to adjust the
  slot number of slotted components. The range is restricted to valid slots
  for the selected component.

* New bitmaps have been added;

      * geda-deslect.png
      * geda-find-attribute
      * geda-properties.png
      * geda-trash-can.png

* libgedathon no longer over-writes existing gafrc when the declare_local_sym
  feature is used.

* libgedathon now supports internationalization (i18n).

Notable changes in gEDA/gaf gEDA 2.1.1.0 - 01/01/17 Rampage Edition

* Improvements were made to both the build system and source code to ensure
  gEDA-GAF builds more reliably on 64 bit platforms. Testing was performed
  using Ubuntu 12.04 and Ubuntu 14.04 64 bit host systems.

* The libgeda API has been completely redone in order to present a library API
  that is not only for in-suite applications.

* Support for internationalization (i18n) has been significantly expanded and
  now includes command-line utilities.

* Extensive revisions have been made to improve performance of gEDA application.
  This is particularly noticeable in gschem.

* A new library, libgedacolor, now unifies color management for the gEDA suite.
  Color-maps are still configured using scheme files, but the use of "friendly-
  color-maps"  is no longer applicable; the new color maps for libgedacolor are
  already "friendly". The library was added to provide a unified API for gEDA
  applications and should be transparent to most users. User who are using
  custom color-maps will need to modify their custom map files.

* A new auto-panning feature has been added to gschem. The auto-pan is only in
  effect while drawing nets and buses and can enabled or disabled through menu
  options.

* The Auto-Number dialog in gschem no longer requires the use of wild-card
  characters in the scope text. This is a new advanced feature that allows
  automatic re-numbering of duplicated reference designators.

* The path for logs file can now be set in initialization files, including
  local gafrc files. This allows logs to be organized on a per project basis.
  The location for logs can also be set using a new GEDALOGS environment variable.

* A new libgeda keyword "log-time" can be use to enable or disable prefixing
  log entries with the time of day.

* The Multi-Attribute Editor Dialog entry now has attribute "word" completion
  identical to the Single-Attribute Editor Dialog.

* The list of valid attribute used by gsymcheck when checking symbol files is
  now users configurable. The list can be set in system-gsymcheckrc.

* The list of known devices recognized by gsymcheck when checking the device
  entries in symbol files is now users configurable. The list can be set in
  system-gsymcheckrc.

* Symbol files can now be renamed using the Component Dialog.

* The number of user defined colors was increased by 6 to a total of
  10 user definable colors in each color map.

* The main menu system has been modified to automatically build a basic
  fall-back menu when errors are discovered in the menu RC file.

* The height of the status bar in Gschem can now be set in the gschem-gtkrc

* HotKey dialog now includes actions icon and a column to display the actual
  binding string. User can enable or disable viewing the action column with
  the setting being retained between session.

* The Slot Editor Dialog now displays the total number of slots per component.

* New features have been added to the Page Select Dialog:

      Accept drag-and-drop of schematic or symbol files.
      Opened pages can be re-ordered either up or down.
      Auto scrolls to selection.
      Shows long filename as tooltip when showing short names.
      Optionally adjust height of dialog automatically.

* Exchange image file now retains the user preference to update all selected
  images between session. Previously the user preference was only retained for
  current session.

* Additional preference variables have been incorporated in new the configuration
  system allow retention of changes using the Preferences dialog:

      continue-component-place
      embed-components
      enforce-hierarchy
      force-boundingbox
      keyboardpan-gain
      netconn-rubberband
      select-slack-pixels
      snap-size
      sort-component-library
      object-clipping

  Uses of these keywords in rc files is still permissible and will over-ride the
  preference setting.

* A new gschem command line options has been provided to specify a title-block at
  startup.

* A new gschem command line options has been provided to startup minimized.

* The command line option to add a directory to the Guile path now has a long option.
  The "-L" option has been replaced with "-g" or "--guile".

* gsymcheck now supports the standard command-line version options.

* Programs in the geda-gaf suite now support quite mode in conjunction with the version
  option to assist with extraction of dotted version information.

* New Scheme API have been added:

      gschem-bmp-image
      gschem-jpeg-image
      gschem-tiff-image

* The maximum zoom magnification in gschem has been increased slightly.


libgedathon/scripts/geda-snap-grid.py
#! /usr/bin/env python

# Set Visibility of Attributes (For now just to Defaults)

import os, sys, glob
from geda import geda
from geda.constants import *

#---------------------------------------------------------------------------
Version="0.4.1"
#---------------------------------------------------------------------------
VerboseMode=False
#---------------------------------------------------------------------------

#  This is the help string.
Usage =\
"""
geda-snap-grid -- A gEDA-gaf Schematics and Symbol cleaning utility

\tRound X-Y coordinates of attributes in schematic and symbol files to
\tmultiples of 25 or a specified value.

\tBy default, geda-snap-grid operates on both schematic and symbol files
\tunless the file names are specified on the command line, either with the
\t--input <inputfile> option or just listing the files. For both schematic
\tand symbol files, the extension must be specified when the names of files
\tare specified.

Usage: geda-snap-grid [Options] -i <inputfile> [[-o] <outputfile> ] [list of files]
"""

Help =\
"""
Options:

  The following options can be used to modify the default behavior of geda-snap-grid

  -r, --recursive --  Process all files in the current and all subordinated directories.
  -v, --verbose   --  Verbose mode, spews lots of info about what the prog is doing.

  -a, --append <path> When processing schematic files this option can be used to specify
                      addition directories to be appended to the libraries search path.
                      This option is not needed for directories with the name "sym", if
                      such a directory exist, the directory will be appended automatically.

  -n, --no-path   --  By default, if a directory exist with the name "sym", then the
                      directory is appended to the libraries search path when processing
                      schematic files, this option inhibits this behavior.

  -c, --sch       --  Only process schematic files, ignored if a file name or names are specified.
  -s  --sym       --  Only process symbol files, ignored if a file name or names are specified.

  -u, --up   Force rounding up, defaults to nearest.
  -d, --down Force rounding down, defaults to nearest.

  -m # Round to moduli specified by #, default is 25.
  -x   Only round X coordinates, mutually exclusive with -y
  -y   Only round Y coordinates, mutually exclusive with -x

Example: Snap all attribute coordinates values in all files in the current
         directory and in all subdirectories to the nearest multiple of 50 and
         report modified values:

    geda-snap-grid.py --verbose -r -m 50

Copyright (C) 2015 by Wiley Edward Hill. Released under GPL Version 2.

"""

#-------------------------------- Classes ----------------------------------
class ProgramParameters:
    """
    This class holds info about the environment and cmd line args passed to the
    program.  It has only one method:  the constructor, which gets the args
    and fills out the public vars.  The public vars are:
    """
    def __init__(self):

        global Version

        """
        Constructor: parse through cmd line args and fill out vars. The values
        here are the defaults
        """
        self.RecursiveMode   = False
        self.VerboseMode     = False
        self.OnlySchematics  = False
        self.OnlySymbols     = False
        self.AutoAppendPath  = True
        self.InputFiles      = []
        self.InputFileName   = ""
        self.OutputFileName  = ""
        self.Path2Symbols    = ""
        self.UserDir         = os.path.abspath(os.getcwd())

        self.rnd_up   = False
        self.rnd_down = False
        self.fix_x    = True
        self.fix_y    = True
        self.moduli   = 25

        if len(sys.argv) > 1:

            mflag = False
            iflag = False
            oflag = False
            pflag = False

            for arg in sys.argv[1:]:       # Skip OurSelf

                if mflag:
                    self.moduli = int(arg)
                    mflag = False
                    continue

                if iflag:
                    self.InputFileName = arg
                    iflag = False
                    continue

                if oflag:
                    self.OutputFileName = arg
                    oflag = False
                    continue

                if pflag:
                    self.Path2Symbols = arg
                    pflag = False
                    continue

                if arg in ('-V', '--version'):
                    print Version
                    sys.exit(0)

                if arg in ('-h', '--help'):
                    print Usage
                    print Help
                    sys.exit(0)

                if arg == '--usage':
                    print Usage
                    sys.exit(0)

                if arg in ('-i', '--input'):
                    iflag = True
                    continue

                if arg in ('-o', '--output'):
                    oflag = True
                    continue

                if arg in ('-a', '--append'):
                    pflag = True
                    continue

                if arg in ('-n', '--no-path'):
                    self.AutoAppendPath = False
                    continue

                if arg in ('-v', '--verbose'):
                    self.VerboseMode = True
                    continue

                if arg in ('-r', '--recursive'):
                    self.RecursiveMode = True
                    continue

                if arg in ('-c', '--sch'):
                    self.OnlySchematics = True
                    continue

                if arg in ('-s', '--sym'):
                    self.OnlySymbols = True
                    continue

                if arg in ('-d', '--down'):
                    self.rnd_down = True
                    continue

                if arg in ('-u', '--up'):
                    self.rnd_up = True
                    continue

                if arg in ('-x', '-X'):
                    self.fix_y = False
                    continue

                if arg in ('-y', '-Y'):
                    self.fix_x = False
                    continue

                if arg in ('-m', '-M', '--moduli'):
                    mflag = True
                    continue

                self.InputFiles.append(arg)

#---------------------------------------------------------------------------
def round_down(num, divisor):
    return num - (num%divisor)
#---------------------------------------------------------------------------
def round_up(num, divisor):
    return num + (num%divisor)
#---------------------------------------------------------------------------

def SnapText(Options, Text):
    modified = False
    if Options.fix_x:
        if not Text.x % Options.moduli == 0:
            modified = modified + 1
            low  = Text.x - Text.x % Options.moduli
            high = Text.x + Options.moduli - Text.x % Options.moduli
            if Options.VerboseMode:
                 print "Text: " + Text.string + " " + str(Text.x) + " is off X grid"
            if Options.rnd_down:
                Text.x = low
            elif Options.rnd_up:
                Text.x = high
            elif Text.x - low <= high - Text.x:
                Text.x = low
            else:
                Text.x = high

    if Options.fix_y:
        if not Text.y % Options.moduli == 0:
            modified = modified + 1
            low  = Text.y - Text.y % Options.moduli
            high = Text.y + Options.moduli - Text.y % Options.moduli
            if Options.VerboseMode:
                 print "Text: " + Text.string + " " + str(Text.y) + " is off Y grid"
            if Options.rnd_down:
                Text.y = low
            elif Options.rnd_up:
                Text.y = high
            elif Text.y - low <= high - Text.y:
                Text.y = low
            else:
                Text.y = high

    if modified and Options.VerboseMode:
        print "New postion: (" + str(Text.x) + "," + str(Text.y) + ")"

    return modified

#---------------------------------------------------------------------------

def ProcessFile(Options, File):

    gedafile = geda.open_page(File)

    if Options.VerboseMode:
        print "Processing: " + gedafile.name()

    objects  = geda.get_objects(gedafile)
    modified = 0
    """
      Loop thru all objects and get list of detached text attribute then loop
      again and skip if in list, Sub object (attrinutes) are to be handled by
      the objects. This is so that detached text gets added to the gedafile
      sym groups.
    """
    for capsule in objects:
        object = geda.get_object(capsule)          # Retrieve object from GedaCapsule
        if capsule.type == OBJ_TEXT:
            if SnapText(Options, object):
                modified = modified + 1

    if modified:
        if Options.OutputFileName:
            OutputFileName = Options.OutputFileName
            gedafile.filename = Options.OutputFileName # Write the file to storage
            print "Using " + OutputFileName + " for output file name"
        else:
            OutputFileName = gedafile.filename

        gedafile.save() # Write the file to storage

        if Options.VerboseMode:
            print "Made " + str(modified) + " modifications to " + OutputFileName

    # Close the file
    gedafile.close()

#---------------------------------------------------------------------------

def RecursiveGlob(path, *exts):
	""" Recursively Glob directory and all subdirectories for multiple file
	extensions. Note: Glob is case-insensitive, i. e. for '\*.sch' you will
	get files ending with .sch and .SCH

    Parameters
    ----------
    path : str
        A directory name
    exts : tuple
        File extensions to glob for

    Returns
    -------
    files : list
        list of files matching extensions in exts in path and subfolders
	"""
	dirs = [a[0] for a in os.walk(path)]
	f_filter = [d+e for d in dirs for e in exts]
	return [f for files in [glob.iglob(files) for files in f_filter] for f in files]

#---------------------------------------------------------------------------
#  Main prog begins here

def main(argv):

	foundfiles = False
	foundOurSelf = False

	Options = ProgramParameters()  #  Creates Options object for program varibles.

	# Resolve library search paths per Options
	if not Options.OnlySymbols:
		if Options.AutoAppendPath:
			if os.path.isdir("sym"):
				append_symbol_path("sym");
		if Options.Path2Symbols:
			if os.path.isdir(Options.Path2Symbols):
				append_symbol_path(Options.Path2Symbols);
			else:
				print Options.Path2Symbols + " does not exist"
				exit(1);

	# Rearrange input list if wild-card was used with -i
	if Options.InputFileName and Options.InputFiles:
		Options.InputFiles.insert(0, Options.InputFileName)
		Options.InputFileName=""

	if Options.InputFileName:
		ProcessFile(Options, Options.InputFileName)
	else:

		# Multi-file mode
		if Options.OutputFileName:
			print "output file can not be specified with multiple input files"

		InputFiles = Options.InputFiles

		files = []

		base = os.getcwd()

		if not InputFiles:
			# Get list of files OurSelf
			foundOurSelf = True
			if Options.OnlySymbols:
				if Options.RecursiveMode:
					files = RecursiveGlob(base, '/*.sym')
				else:
					files = glob.glob(os.path.join(base,'*.sym'))
			elif Options.OnlySchematics:
				if Options.RecursiveMode:
					files = RecursiveGlob(base, '/*.sch')
				else:
					files = glob.glob(os.path.join(base,'*.sch'))
			elif Options.RecursiveMode:
				files = RecursiveGlob(base,'/*.sym', '/*.sch')
			else:
				files = glob.glob(os.path.join(base,'*.sym'))
				files.extend(glob.glob(os.path.join(base,'*.sch')))

		if files:
			foundfiles=True
			for file in files:
				InputFiles.append(file)

		if foundOurSelf and len(InputFiles) == 0:
			print 'No symbols found or specifed, try using --help'

		elif not foundfiles :
			if len(InputFiles) > 1:
				for file in InputFiles:
					if not os.path.isfile(file):
						print "Bad file name or option: " + file
						print Usage
		else:
			for symbol in InputFiles:
				ProcessFile(Options, symbol)

if __name__ == "__main__":
   main(sys.argv[1:])

# Done!
